FIX _safe_indexing for pyarrow

lorentzenchr opened the pull request:
#### Reference Issues/PRs
Partially addresses https://github.com/scikit-learn/scikit-learn/issues/25896#issuecomment-2740201980.

#### What does this implement/fix? Explain your changes.
`_safe_indexing(.., axis=1)` is used in the `ColumnTransformer` and raises an error if a `pyarrow.Table` is passed even though it implements the __dataframe__ interchange protocol:
```python
import pyarrow as pa
from sklearn.datasets import load_iris
from sklearn.preprocessing import StandardScaler, KBinsDiscretizer
from sklearn.compose import ColumnTransformer

X, y = load_iris(as_frame=True, return_X_y=True)
sepal_cols = ["sepal length (cm)", "sepal width (cm)"]
petal_cols = ["petal length (cm)", "petal width (cm)"]

preprocessor = ColumnTransformer(
    [
        ("scaler", StandardScaler(), sepal_cols),
        ("kbin", KBinsDiscretizer(encode="ordinal"), petal_cols),
    ],
    verbose_feature_names_out=False,
)

X_pa = pa.table(X)
preprocessor.fit_transform(X_pa)
```
results in 
```
python3.10/site-packages/sklearn/utils/_indexing.py:270, in _safe_indexing(X, indices, axis)
    268     return _polars_indexing(X, indices, indices_dtype, axis=axis)
    269 elif hasattr(X, "shape"):
--> 270     return _array_indexing(X, indices, indices_dtype, axis=axis)
    271 else:
    272     return _list_indexing(X, indices, indices_dtype)

File python3.10/site-packages/sklearn/utils/_indexing.py:36, in _array_indexing(array, key, key_dtype, axis)
     34 if isinstance(key, tuple):
     35     key = list(key)
---> 36 return array[key, ...] if axis == 0 else array[:, key]

...
python3.10/site-packages/pyarrow/table.pxi:1725, in pyarrow.lib._Tabular._ensure_integer_index()

TypeError: Index must either be string or integer
```
which shows that the wrong branch (`elif hasattr(X, "shape"):`) is taken.


#### Any other comments?
There is no general solution with `__dataframe__` because of https://github.com/data-apis/dataframe-api/issues/85.

~~Therefore a dirtier solution is taken.~~
Therefore, just pyarrow indexing is implemented.

[narwhals](https://github.com/narwhals-dev/narwhals) would be much cleaner, but needs its own dedicated issue for discussion. This PR just fixes a bug with pyarrow.Table passed around.

lorentzenchr committed:
FIX _safe_indexing for __dataframe__ interchange protocol

ogrisel reviewed (approved):
+1 for supporting arrow, and I fine with doing it this way as a pragmatic and incremental improvement pending a potential refactoring to make it more general in the future (to be discussed in a follow-up issue or PR).

However I am worried that this code is untested on our CI. Could you add PyArrow as a soft dependencies on one of the worker CIs (like we do for `pandas` / `polars`/ `pytorch` / `array-api-strict`...)?

lorentzenchr commented:
> However I am worried that this code is untested on our CI.

I don't follow. `scikit-learn/build_tools/azure/pylatest_conda_forge_mkl_linux-64_environment.yml` and `pylatest_conda_forge_cuda_array-api_linux-64_environment.yml` both specify `pyarrow` as requirement, so the code path of this PR is tested.

lorentzenchr commented:
Should the changelog be in `sklearn.compose` for the `ColumnTransformer` or in `sklearn.utils` for `_safe_indexing`?

lorentzenchr committed:
DOC add whatsnew entry

ogrisel commented:
> Maybe our codevoc is not clever enough (configured).

Actually, everything is fine: the coverage data is collected on the CI config that runs with `pylatest_conda_forge_mkl_linux-64_environment.yml` and pyarrow. It's just that we actually do not cover the lines reported by codecov in the tests. For instance the `NotImplementeError` case is trick to cover.

lorentzenchr commented:
@ogrisel Do we need a 2nd reviewer?

jeremiedbb commented on sklearn/utils/_indexing.py:
```diff
@@ -93,6 +93,40 @@ def _polars_indexing(X, key, key_dtype, axis):
+        )
+        raise NotImplementedError(msg)
+
+    if key_dtype == "bool":
+        key = np.asarray(key).nonzero().tolist()
+        key_dtype == "int"
```
this is not covered. I think a test should easily doable for it

jeremiedbb commented on sklearn/utils/_indexing.py:
```diff
@@ -93,6 +93,40 @@ def _polars_indexing(X, key, key_dtype, axis):
+    # https://github.com/data-apis/dataframe-api/issues/85.
+    # As a dirty workaround, we will check for a method "select" being available and
+    # use that one.
+    if not hasattr(X, "select"):
```
Okay to no cover this so let's be explicit about it
```suggestion
    if not hasattr(X, "select"):  # pragma: no cover
```

jeremiedbb commented on sklearn/utils/_indexing.py:
```diff
@@ -93,6 +93,40 @@ def _polars_indexing(X, key, key_dtype, axis):
+
+    if isinstance(key, np.ndarray):
+        key = key.tolist()
+    elif np.isscalar(key):
+        key = list(key)
```
This is not covered, I also think we should add a test for it

jeremiedbb commented on sklearn/utils/tests/test_indexing.py:
```diff
@@ -172,7 +174,11 @@ def test_safe_indexing_2d_container_axis_1(array_type, indices_type, indices):
     indices_converted = _convert_container(indices_converted, indices_type)
 
-    if isinstance(indices[0], str) and array_type not in ("dataframe", "polars"):
+    if isinstance(indices[0], str) and array_type not in (
+        "dataframe",
+        "pyarrow",
+        "polars",
+    ):
```
nitpick
```suggestion
    if isinstance(indices[0], str) and array_type in ("array", "sparse"):
```

jeremiedbb reviewed (commented):
Hey, given that there seems to be a consensus to use narwhals for our dataframe support tools, do you think that this PR is still relevant ?
We can easily fit it in 1.7 while it's unlikely for narwhals. But do we want to introduce this is we're going to rewrite it soon (ref https://github.com/scikit-learn/scikit-learn/issues/31049) ?

lorentzenchr commented:
> do you think that this PR is still relevant ?

Yes, it’s a simple fix for a bug. If we later decide to go with narwhals, there is no cost of having this PR.

lorentzenchr committed:
CLN address review comments

lorentzenchr committed:
ENH _pyarrow_indexing replaces _dataframe_interchange_protocol_indexing

lorentzenchr committed:
DOC change whatsnew to enhancement in modul utils

lorentzenchr commented:
@ogrisel @jeremiedbb I still would like to have this PR in 1.7. But I changed my mind about the implementation. I replaced the "dataframe interchange protocol" indexing by proper pyarrow indexing. This is much cleaner.

I'm now convinced that the dataframe interchange protocol is not usable for anything else but interchange which we don't really do or need, we need an API for indexing and assigning. Therefore, I defer this topic to #31049.

ogrisel reviewed (approved):
LGTM.

jeremiedbb commented on sklearn/utils/tests/test_indexing.py:
```diff
@@ -187,14 +191,18 @@ def test_safe_indexing_2d_container_axis_1(array_type, indices_type, indices):
 def test_safe_indexing_2d_read_only_axis_1(
     array_read_only, indices_read_only, array_type, indices_type, axis, expected_array
 ):
+    if array_type == "pyarrow" and axis == 0:
+        pytest.skip()
```
I don't think we need to skip since you implemented axis-0 indexing for pyarrow

jeremiedbb reviewed (approved):
LGTM. Just 1 small thing left

lorentzenchr committed:
MNT remove skip in test_safe_indexing_2d_read_only_axis_1
